### DISCUSSION TITLE
Dynamic select input fields change value after change when they should not

### DISCUSSION THREAD

**Discussion by ojsoft** [livewire/livewire#6914] [7. Bugs (archived)]
### Livewire version

3.0.5

### Which PHP version are you using?

PHP 8.1

### Steps To Reproduce

1. Make a dynamic `<select wire:model.live="test">` field. Dynamic means that the number of options inside the select get changed by another property.
2. Change the other property so that select options get stripped out.
3. As you will see the Select field will keep the position of the select, but not the value.

### Example

__Laravel component__ contains
```
public $greys = true;
public $color = 'red';
```

__Blade__ contains
```
<select wire:model.live="color">
   @if($greys)
      <option value="white">White</option>
      <option value="black">Black</option>
      <option value="grey">Grey</option>
   @endif
   <option value="red">Red</option>
   <option value="green">Green</option>
   <option value="blue">Blue</option>
   <option value="orange">Orange</option>
</select>

<input type="checkbox" wire:model.live="greys"/> Allow Greys

```

Ok, let's assume you switch the Greys checkbox off. What then happens is that the color implicitly changes from Red to Orange. So position 4 in select is kept, not the actual value of red.

I make highly reactive forms with Laravel and Livewire 2 did not have this problem. So I guess this must be a bug in livewire 3.

Just figured out there is another problem in the same hemisphere.

__Laravel component__ contains
```
public $color = 'red';

public function save()
{
     // $this->color is always red when the form is submitted if the user did not actively change the select to any color
}
```

__Blade__ contains
```
<form wire:submit="save">
   <select wire:model.live="color">
      <option value="green">Green</option>
      <option value="blue">Blue</option>
      <option value="orange">Orange</option>
   </select>
   <button>Send</button>
</form>
```

When this form is sent, the user sees Green in the select, but the value green is not changed by the form, the value of red stays in the component. This also worked nicely in Livewire 2 and is a new surprising "feature" of Livewire 3. You can argue that the mistake is that the color is set to a value that does not exist inside the form. But... with such dynamic selects it crucially changes the expected behavior. When a html form is sent, it sends the first value in the list as the new value. This is established behavior and tons of my code rely on this. Could this please be restored.

I also found out that $set('XXX', 1) calls do not reach the updatedXXX function when XXX already has the value of 1. My code sadly relied on this too. In that case I can understand that it is cleaner this way, but a big red marker in the upgrade documentation on this changed behavior would have been nice.

### Additional information

Sorry, just realized that my markdown had a problem. Fixed it in edit. Now it is all well readable.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by kevin-r-j** [comment]
I came across this same probelm and agree it is a bug but you can fix it by dynamically declaring the selected attirbute on the option tag. That will take precedence over any guesswork that Livewire or Alpine are assuming.

Even on a small set of options, I usually set an options array in the class public attirbutes and loop through them to create each option. That means I only have to add the condition once
`@foreach($colorOptions as $colorOption)`
     `    <option @if($color === $colorOption) selected @endif value="{{ $colorOption }}">{{ $colorOption }}</option>`
`@endforeach`

Regarding the second issue, this is because the select has no alternative but to display one of the available colors. You need to add an empty option (and required attribute on the select if it suits) so that if your initial `$color` is not in the list it will not automatically present another color.

**Reply by jp-guiang** [reply]
We have this issue as well, the closest we got to a solution was to add "wire:key={{ $color_id }}" but you still need to include the logic check to confirm that the "selectedColor" is matching the "color_id".

Which really means that the original problem stands, and specifically that in Livewire 2, "wire:model" would dynamically update dropdown lists that are related to each other, in blade files. But this no longer works in Livewire 3, and the breaking change is not documented.

An interesting behaviour that we saw was when we type inspected the Eloquent model keys/ids that were being passed around, some were being sent as strings and others as integers.

**Reply by peterjthomson** [reply]
We managed to find a passable workaround using `@selected($color === $colorOption)` blade helper inside the <option > list items. But it feels like we are undoing the point of using wire:model on the list in the first place.
